src/middlewared/middlewared/api/v25_04_0/nfs.py

```diff
@@ -0,0 +1,194 @@
+import re
+from ipaddress import ip_network
+from typing import Annotated, Literal
+
+from pydantic import (
+    ConfigDict, Field, IPvAnyAddress, IPvAnyNetwork,
+    AfterValidator, field_validator
+)
+
+from middlewared.api.base import (
+    BaseModel, Excluded, excluded_field, ForUpdateMetaclass, NonEmptyString,
+    single_argument_args, single_argument_result,
+    TcpPort, exclude_tcp_ports
+)
+
+__all__ = ["NfsEntry",
+           "NfsUpdateArgs", "NfsUpdateResult",
+           "NfsBindipChoicesArgs", "NfsBindipChoicesResult",
+           "NfsShareEntry",
+           "NfsShareCreateArgs", "NfsShareCreateResult",
+           "NfsShareUpdateArgs", "NfsShareUpdateResult",
+           "NfsShareDeleteArgs", "NfsShareDeleteResult"]
+
+NFS_protocols = Literal["NFSV3", "NFSV4"]
+NFS_RDMA_DEFAULT_PORT = 20049
+EXCLUDED_PORTS = [NFS_RDMA_DEFAULT_PORT]
+
+
+class NfsEntry(BaseModel):
+    id: int
+    servers: Annotated[int | None, Field(ge=1, le=256)]
+    """ Specify the number of nfsd. Default: Number of nfsd is equal number of CPU. """
+    allow_nonroot: bool
+    """ Allow non-root mount requests.  This equates to 'insecure' share option. """
+    protocols: list[NFS_protocols]
+    """ Specify supported NFS protocols:  NFSv3, NFSv4 or both can be listed. """
+    v4_krb: bool
+    """ Force Kerberos authentication on NFS shares. """
+    v4_domain: str
+    """ Specify a DNS domain (NFSv4 only) """
+    bindip: list[NonEmptyString] = Field(default_factory=[])  # Using IPvAnyAddress creates JSON serialization errors
+    """ Limit the server IP addresses available for NFS """
+    mountd_port: Annotated[TcpPort | None, AfterValidator(exclude_tcp_ports(EXCLUDED_PORTS))]
+    """ Specify the mountd port binding """
+    rpcstatd_port: Annotated[TcpPort | None, AfterValidator(exclude_tcp_ports(EXCLUDED_PORTS))]
+    """ Specify the rpc.statd port binding """
+    rpclockd_port: Annotated[TcpPort | None, AfterValidator(exclude_tcp_ports(EXCLUDED_PORTS))]
+    """ Specify the rpc.lockd port binding """
+    mountd_log: bool
+    """ Enable or disable mountd logging """
+    statd_lockd_log: bool
+    """ Enable or disable statd and lockd logging """
+    v4_krb_enabled: bool
+    """ Status of NFSv4 authentication requirement (status only) """
+    userd_manage_gids: bool
+    """ Enable to allow server to manage gids """
+    keytab_has_nfs_spn: bool
+    """ Report status of NFS Principal Name in keytab (status only)"""
+    managed_nfsd: bool
+    """ Report status of 'servers' field.
+    If True the number of nfsd are managed by the server. (status only)"""
+    rdma: bool
+    """ Enable or disable NFS over RDMA.  Requires RDMA capable NIC """
+
+    @field_validator('bindip')
+    @classmethod
+    def check_bind_ip(cls, field_value: list):
+        """ Custom validator for IP addresses to avoid JSON serialization errors """
+        all(isinstance(v, IPvAnyAddress) for v in field_value)
+        return field_value
```

Comment:
This doesn't seem right for a few reasons:
- result of `all` is not being tested
- Don't we need to **instantiate** `IPvAnyAddress(v)` to see if it'll throw an error?

Maybe something like:
```
    def check_bind_ip(cls, field_value: list):
        """ Custom validator for IP addresses to avoid JSON serialization errors """
        for v in field_value:
            IPvAnyAddress(v)
        return field_value
```